### ISSUE TITLE
A bug that can be used to duplicate any items!

### ISSUE THREAD

**Issue by azbh111** [AppliedEnergistics/Applied-Energistics-2#2655]
My video:
https://youtu.be/bCB58sHRXxU
or
https://www.youtube.com/watch?v=bCB58sHRXxU&feature=youtu.be

**Comment by mindforger** [comment]
and which AE2 Version did you use ? which minecraft version is that?

**Comment by azbh111** [comment]
appliedenergistics2-rv3-beta-6

**Comment by azbh111** [comment]
forge:1.7.10-10.13.4.1614-1.7.10

**Comment by mindforger** [comment]
oookay, i am gonna try this on evening XD

**Comment by azbh111** [comment]
now I have to ban the ME Controller    
QAQ

**Comment by mindforger** [comment]
do you have channels disabled on your video ?

**Comment by azbh111** [comment]
what did you mean?

**Comment by azbh111** [comment]
= =

**Comment by azbh111** [comment]
I test with only AE2mod,the bug is still exist.
and rv2-stable-10 has the same bug

**Comment by azbh111** [comment]
I have other ideas but ban the controller

**Comment by azbh111** [comment]
Over the past few months, some players duplicated a lot of items.
I know it,but I know the method today.
I think it related  to the ME system which contains ME controller

**Comment by mindforger** [comment]
i will test this in a few hours if not somebody else is faster, stay a bit patient, i am sure it is only a special case and if not it will work with similar setups without controller also

**Comment by azbh111** [comment]
thanks  :)

**Comment by azbh111** [comment]
multiplayer server: Thermos-1.7.10-1614-58-server
https://tcpr.ca/downloads/thermos
 single player:forge:1.7.10-10.13.4.1614-1.7.10
I tested on both single and multiplayer and both were successful

**Comment by mindforger** [comment]
yeah i can confirm this @yueh and i also have a strong guess

you have to place a cable and break it just befor the me drive goes online 

i presume the breaking of the cable triggers an update of the storage like it does when breaking the cable but with a false inventory set

**Comment by yueh** [comment]
Even if it should happen, as long as it does not affect rv4 we will not fix it.

**Comment by mindforger** [comment]
works fine on 1.10.2 rv4 a6

**Comment by yueh** [comment]
I think I have fixed it, will do a PR in a few minutes to test (for rv4)

**Comment by yueh** [comment]
@mindforger Could you please fetch the build for #2657 from jenkins and verify it?

**Comment by mindforger** [comment]
in about 10hours or so i can test :)

**Comment by mindforger** [comment]
is there a slight chance this will turn into an rv3 a7 eventually ? :) if by chance this video/info spreads it will ruin a lot of 1.7.10 servers i guess

**Comment by azbh111** [comment]
I have deleted this video.  :)

**Comment by mindforger** [comment]
removed my explanation for the sake of running servers :)

**Comment by yueh** [comment]
It is actually a duplicate of #2276, so there are already enough documented ways to reproduce it.

And no we will not release a new rv3 build. Potentially a stable to preserve the last version as moving rv4 to beta will overwrite them.

**Comment by mindforger** [comment]
that one O_o i thought this one was solved already ... well okay then :) 
wait, hope and ban everybody who misused it :P

**Comment by azbh111** [comment]
I saw #2276 ,then I found that two separate ME systems can be confusing.
Wait for my video

**Comment by azbh111** [comment]
Even if fixed #2657, I can still duplicate  items with #2276.

**Comment by mindforger** [comment]
not if it is related, it's most likely the same bug and the fix would help against both probably

**Comment by azbh111** [comment]
After my test,it do not fix both.

**Comment by azbh111** [comment]
//fix
this.inactiveCellProviders.remove(cc);      //++
this.getGrid().postEvent(new MENetworkCellArrayUpdate());
this.removeCellProvider(cc, new CellChangeTracker()).applyChanges();
//this.inactiveCellProviders.remove(cc);      //--

I build GridStorageCache.class myself.
#2655 fixed,but #2276 not
I'm recording a video for #2276

**Comment by mindforger** [comment]
ah okay sorry i misunderstood your post the first time, you already tested it, thank you :)

**Comment by azbh111** [comment]
https://youtu.be/aqFz9YLdHgA
appliedenergistics2-rv3-beta-6  (#2655 has been fiexd)
forge:1.7.10-10.13.4.1614-1.7.10
single player
I am not sure this video is detailed enough for you.
If not,tell me.

**Comment by mindforger** [comment]
so it seems like the cache is not updated properly while somebody is looking at the interface

**Comment by azbh111** [comment]
I am waiting for the fixing code
:)

**Comment by azbh111** [comment]
Has https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/fc834036a0a76daec3cf7d39bad383d97a11ac0d/src/main/java/appeng/me/NetworkEventBus.java#L152  fixed #2276 ?

**Comment by mindforger** [comment]
i changed the line in my comment, i was mistaken for the call, it is line 126 not 152 and no it is not fixed

but i pointed out that this empty catch may be dropping an unhandled exception, you can try to insert a log output there and see if it fires when you make that exploit

**Comment by azbh111** [comment]
there is no exception be thrown.

**Comment by azbh111** [comment]
MENetworkEvent postEvent(final Grid g, final MENetworkEvent e) {
        final Map<Class, MENetworkEventInfo> subscribers = EVENTS.get(e.getClass());
        int x = 0;

        try {
            if (subscribers != null) {
                for (final Entry<Class, MENetworkEventInfo> subscriber : subscribers.entrySet()) {
                    final MENetworkEventInfo target = subscriber.getValue();
                    final GridCacheWrapper cache = g.getCaches().get(subscriber.getKey());
                    if (cache != null) {
                        x++;
                        target.invoke(cache.getCache(), e);
                    }

                    for (final IGridNode obj : g.getMachines(subscriber.getKey())) {
                        x++;
                        target.invoke(obj.getMachine(), e);
                    }
                }
            }
        } catch (final NetworkEventDone done) {
            // Early out.
            System.out.println("================================");
            System.out.println("126");
            System.out.println("================================");
        }

        e.setVisitedObjects(x);
        return e;
    }

    MENetworkEvent postEventTo(final Grid grid, final GridNode node, final MENetworkEvent e) {
        final Map<Class, MENetworkEventInfo> subscribers = EVENTS.get(e.getClass());
        int x = 0;

        try {
            if (subscribers != null) {
                final MENetworkEventInfo target = subscribers.get(node.getMachineClass());
                if (target != null) {
                    x++;
                    target.invoke(node.getMachine(), e);
                }
            }
        } catch (final NetworkEventDone done) {
            System.out.println("================================");
            System.out.println("152");
            System.out.println("================================");
        }
        e.setVisitedObjects(x);
        return e;
    }

**Comment by mindforger** [comment]
darn it, okay i was looking for the connection between an open terminal view and the broken cable 

but i only know about the remove() call for a node and this was the only connection betwen them

**Comment by azbh111** [comment]
during testing #2276 ,I happen to found that two separate ME systems can be confusing.
I will study how to reproduce it tomorrow.

**Comment by yueh** [comment]
It is funny how this takes the exact same route as the last one..

1) I certainly will not watch a video longer than maybe 20-30 seconds without any explanation. There is simply too much stuff going on to track, consider important and try to remember it without having a chance to go back and reread a certain section. Compared to a written report with additional screenshots or an actual world download.

2) Stop guessing that completely irrelevant things are responsible. Yes the `try {} catch {}` to handle cancled events might not be ideal compared to a `if (event.isCanceled) return;`. But it works and nothing in AE2 ever cancels these events. I can either spent my time fixing stuff or writing down why the proposed solution is useless.

Nevertheless, should be fixed now once the new build for the PR is done.

**Comment by mindforger** [comment]
whoops forgot to write last evening
i can confirm that this issue with placebreaking a cable is FIXED
but the referenced #2276 is still NOT FIXED and NOT YET REPRODUCED with rv4 only rv3 has been reproduced!

how to reproduce:

build a similar setup like before, open the terminal
let now a seconds player, turtle or blockbreaker rmeove the connection to the drive
the terminal will not update and you can still take out the items from the drive without modifying the contents of the yet disconnected drive!

**Comment by azbh111** [comment]
https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/fix-2655/src/main/java/appeng/me/cache/GridStorageCache.java
I test with it,#2276 is fixed,but #2657 is  reproduced.     = =

**Comment by bookerthegeek** [comment]
@yeah did you want me to test this and the other linked bug on the latest build for you? If so, I'll need steps to reproduce this bug, but the other one I still actually have that save. Lol.

**Comment by mindforger** [comment]
@bookerthegeek make a simple setup:

[drive]  [cable 1] [cable 2 + terminal (add a cable and a storage monitor for the dupe item above)] [controller] [powercell]

now place a storage cell in the drive and drop an item into the terminal

break cable 1, observer storage monitor -> should update immediately to 0

now place the cable 1-> update should take a while since the network is rebuilt

now break the cabel 1 again to prepare the initial glitch 

to execute glitch, place and immediately break the cable again -> the monitor would falsely update on breaking the cable with the items stored in the drive without beeing attached

you could then remove the cell from the drive and take out the items and you have duplicated them

second setup is similar but you need someone or something to break cable 1 WHILE you are in the terminal!

OP  did it with a delayline of repeaters that activate an annihilation pane destroying the cable, he flicked the lever, opened the terminal and the delayed destruction did the trick then, you could also take out all items stored on the cell without modifying it's content because it is not conected, effectively duping everything

**Comment by bookerthegeek** [comment]
Thanks. I'll get some testing setups tonight and report back.

**Comment by azbh111** [comment]
both were fiexd

**Comment by bookerthegeek** [comment]
Nice, I'll test the other bugs then. 😎
